src/shm_main.c

```diff
@@ -768,6 +813,273 @@ sr_shmmain_createunlock(int shm_lock)
+ * @brief Release the specified connection ID.
+ * Free all resources used in tracking the connection.
+ *
+ * @param[in] cid Connection ID to release.
+ * @return err_info, NULL on success.
+ */
+sr_error_info_t *
+sr_shmmain_deallocate_conn(const sr_cid_t cid) {
+    sr_error_info_t *err_info = NULL;
+    char *path = NULL;
+    conn_list_entry *ptr = NULL;
+    conn_list_entry *prev = NULL;
+    conn_list_entry *removed = NULL;
+
+    /* Search the list of active connections owned by this process in order to
+     * remove the connection from the list and clean up */
+    if (NULL != conn_head) {
+        if ( (err_info = sr_mlock(conn_list_lock, 1000, __func__))) {
+            return err_info;
+        }
+
+        for (ptr = conn_head; ptr != NULL; ) {
+            if (cid == ptr->cid) {
+                /* Remove the entry from the list */
+                if (NULL == prev) {
+                    conn_head = ptr->_next;
+                } else {
+                    prev->_next = ptr->_next;
+                }
+                /* Cleanup local resources */
+                if (ptr->cid_fd > 0) {
+                    /* Note that closing ANY file descriptor to a locked file
+                     * releases all locks */
+                    close(ptr->cid_fd);
+                    ptr->cid_fd = 0;
+                } else {
+                    SR_LOG_WRN("Connection CID %d fd is unset", ptr->cid);
+                }
+                /* Removed entry does not adjust prev */
+                /* prev = prev; */
+                removed = ptr;
+                ptr = ptr->_next;
+                free(removed);
+            } else {
+                prev = ptr;
+                ptr = ptr->_next;
+            }
+        }
+        sr_munlock(conn_list_lock);
+    }
+
+    sr_shmmain_get_conn_lockfile_path(cid, &path);
+    if (0 != unlink(path)) {
+        SR_ERRINFO_SYSERRNO(&err_info, "unlink");
+    }
+    free(path);
+
+    return err_info;
+}
+
+/**
+ * @brief Allocate and track a globally unique connection ID.
+ * Populate the sr_cid field of conn with newly allocated ID.
+ * Main SHM write lock is expected to be held.
+ *
+ * @param[in,out] Connection for which an ID will be allocated.
+ * @return err_info, NULL on success.
+ */
+sr_error_info_t *
+sr_shmmain_allocate_conn(sr_conn_ctx_t *conn) {
+    sr_error_info_t *err_info = NULL;
+    char buf[64] = {0};
+    char *path = NULL;
+    struct flock fl;
+    conn_list_entry *cid_info = NULL;
+    int lock_fd = 0;
+    char *owner = NULL;
+    char *group = NULL;
+    mode_t perm = 0;
+    sr_main_shm_t *main_shm = NULL;
+    sr_cid_t cid = SR_CONN_ID_NONE;
+    int alive = 0;
+
+    main_shm = (sr_main_shm_t *)conn->main_shm.addr;
+
+    conn->sr_cid = SR_CONN_ID_NONE;
+    /* Allocate next unique Connection ID */
+    cid = ATOMIC_INC_RELAXED(main_shm->new_sr_cid);
+    err_info = sr_shmmain_check_conn_lock(cid, &alive);
+    if (err_info || (0 != alive) || (SR_CONN_ID_NONE == cid) ) {
```

Comment:
I believe `alive` is technically a boolean except the actual type is not used (which is correct) but please work with it as boolean meaning test it as just `alive` instead of `0 != alive`.